core/src/main/java/com/linecorp/armeria/server/AbstractBindingBuilder.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.linecorp.armeria.common.HttpMethod.CONNECT;
+import static com.linecorp.armeria.common.HttpMethod.DELETE;
+import static com.linecorp.armeria.common.HttpMethod.GET;
+import static com.linecorp.armeria.common.HttpMethod.HEAD;
+import static com.linecorp.armeria.common.HttpMethod.OPTIONS;
+import static com.linecorp.armeria.common.HttpMethod.PATCH;
+import static com.linecorp.armeria.common.HttpMethod.POST;
+import static com.linecorp.armeria.common.HttpMethod.PUT;
+import static com.linecorp.armeria.common.HttpMethod.TRACE;
+import static com.linecorp.armeria.server.HttpHeaderUtil.ensureUniqueMediaTypes;
+import static java.util.Objects.requireNonNull;
+
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.MediaType;
+
+/**
+ * An abstract builder class for binding a {@link Route} fluently.
+ *
+ * @see VirtualHostDecoratingServiceBindingBuilder
+ * @see ServiceBindingBuilder
+ * @see VirtualHostServiceBindingBuilder
+ */
+abstract class AbstractBindingBuilder {
+
+    private Set<HttpMethod> methods = HttpMethod.knownMethods();
+    private Set<MediaType> consumeTypes = ImmutableSet.of();
+    private Set<MediaType> produceTypes = ImmutableSet.of();
+    private final Map<RouteBuilder, Set<HttpMethod>> routeBuilders = new LinkedHashMap<>();
+    private final Set<RouteBuilder> pathBuilders = new HashSet<>();
```

Comment:
I wonder if we need to retain the order of insertion here. If so, we could use `ImmutableSet.Builder` here (and for `routeBuilders` as well for consistency).

Reply:
I fixed `pathaBuilders`, but `routeBuilders` use `computeIfAbsent`.
https://github.com/line/armeria/blob/671f5458e7a6c9a4c3fd7a845f2b0d3f0eed381e/core/src/main/java/com/linecorp/armeria/server/AbstractBindingBuilder.java#L183